### PR TITLE
[Core] update protobuf to 3.19.4

### LINKED ISSUE #25638: [Core] Ray broken when the latest protobuf is used  (>= 4.21.1)
### What happened + What you expected to happen

Seems like Ray will be broken when protobuf >= 4.21.1 is used because it requires building protobuf using the version >= 3.19, but we are using protobuf 3.16 for our core dependency. https://github.com/ray-project/ray/blob/2614c24e475be13d4d67997edd35c19f53f6cb2c/bazel/ray_deps_setup.bzl#L91

```
  File "<stdin>", line 1, in <module>
  File "/Users/saitcakmak/opt/anaconda3/envs/py37/lib/python3.7/site-packages/ray/__init__.py", line 108, in <module>
    import ray._raylet  # noqa: E402
  File "python/ray/_raylet.pyx", line 115, in init ray._raylet
  File "/Users/saitcakmak/opt/anaconda3/envs/py37/lib/python3.7/site-packages/ray/exceptions.py", line 7, in <module>
    from ray.core.generated.common_pb2 import RayException, Language, PYTHON
  File "/Users/saitcakmak/opt/anaconda3/envs/py37/lib/python3.7/site-packages/ray/core/generated/common_pb2.py", line 15, in <module>
    from . import runtime_env_common_pb2 as src_dot_ray_dot_protobuf_dot_runtime__env__common__pb2
  File "/Users/saitcakmak/opt/anaconda3/envs/py37/lib/python3.7/site-packages/ray/core/generated/runtime_env_common_pb2.py", line 42, in <module>
    serialized_options=None, json_name='packages', file=DESCRIPTOR,  create_key=_descriptor._internal_create_key),
  File "/Users/saitcakmak/opt/anaconda3/envs/py37/lib/python3.7/site-packages/google/protobuf/descriptor.py", line 560, in __new__
    _message.Message._CheckCalledFromGeneratedFile()
TypeError: Descriptors cannot not be created directly.
If this call came from a _pb2.py file, your generated code is out of date and must be regenerated with protoc >= 3.19.0.
If you cannot immediately regenerate your protos, some other possible workarounds are:
 1. Downgrade the protobuf package to 3.20.x or lower.
 2. Set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python (but this will use pure-Python parsing and will be much slower).

More information: https://developers.google.com/protocol-buffers/docs/news/2022-05-06#python-updates
```

We fixed the issue by restricting the upper bound of the protobuf version https://github.com/ray-project/ray/issues/25282. However, this can still trigger the same issue when protobuf >= 4.21.1 is downloaded unexpectedly. For example,

```
pip install ray[default]
# Install protobuf >= 4.21.1 by mistake either directly or transitively
```

Ideally, we should just make it work when protobuf >= 4.21.1 is used. Based on the error message, it seems like upgrading the core protobuf version will just fix the issue (there could be other incompatibility issues, although I doubt there will be). 

### Versions / Dependencies

Master

### Reproduction script

N/A

### Issue Severity

_No response_

### PR DESCRIPTION
**Pull request by scv119**
## Why are these changes needed?
The error message in https://github.com/ray-project/ray/issues/25638 indicates we should use protobuf>3.19.0 to generated code so that we can work with python protobuf >= 4.21.1. Try generating  wheels to see if this works.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [x] Manually test
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by scv119** [comment]
Manually tested with this we can work with both protobuf == 4.21.1 and protobuf == 3.20.1

File: python/ray/tests/test_protobuf_compatibility.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    ray.init()
+
+    @ray.remote
+    def load_ray_tune():
+        import google.protobuf
+
+        # verfiy this no longer crashes. see
+        # https://github.com/ray-project/ray/issues/25282
+        from ray import tune  # noqa
```
Comment (matthewdeng) [inline python/ray/tests/test_protobuf_compatibility.py:17]: To simplify this even more, this should fail with just `import ray` right?

**Comment by pcmoritz** [comment]
Thanks a ton for fixing this, this is great! Do we need the upper bound on 4.21.1 in the requirements.txt? It would be great to not constrain it so the CI will detect if there is other versions in the future that cause problems :)

File: python/requirements.txt [hunk 2]
```diff
@@ -18,7 +18,7 @@ msgpack >= 1.0.0, < 2.0.0
 numpy >= 1.16
 opencensus
 prometheus_client >= 0.7.1, < 0.14.0
-protobuf >= 3.8.0, < 4.0.0
+protobuf >= 3.12.0, <= 4.21.1
```
Comment (rkooo567) [inline python/requirements.txt:21]: Do we need upper bound? I think it is safe to just remove it?

Comment (rkooo567) [inline python/requirements.txt:21]: Is it not going to work with protobuf 3.8?

**Review by rkooo567 — APPROVED** [review]
LGMT if tests pass!

File: python/requirements.txt (hunk 2 above)
Comment (scv119) [reply python/requirements.txt:21]: no it's not as i test it.

**Comment by scv119** [comment]
well, we are hitting a windows bug: https://github.com/protocolbuffers/protobuf/issues/10075

**Comment by scv119** [comment]
@pcmoritz @rkooo567  so basically this works for linux and mac... except for Windows. 
Turns out there is an open issue https://github.com/protocolbuffers/protobuf/issues/10075 that python protobuf-4.21.1 doesn't work with `protobuf to 3.19.4` generated python file, on Windows.

What's our option here?
1. we upgrade C++ protobuf to 3.19.4, change the pip requirement to **protobuf >= 3.12.0, < 4.0.0.** In this way:
  - by default, all user should install a protobuf version that's compatible (between 3.12.0 to 4).
  - linux/mac user with protobuf between 3.12.0 to 4.21.1 (latest) will work.
  - windows user, if their installation didn't respect our requirement,  with protobuf 4.21.1 doesn't work.
2. we upgrade C++ protobuf to 3.19.4, change the pip requirement to **protobuf >= 3.12.0 **. In this way:
  - linux/mac user with protobuf between 3.12.0 to 4.21.1 (latest) will work.
  - windows user with latest protobuf 4.21.1, ray doesn't work.
3. do nothing?

**Comment by pcmoritz** [comment]
There is also 4., which is upgrade C++ protobuf to 3.19.4 only on non-Windows (and also unpin), and keep the old one on Windows until https://github.com/protocolbuffers/protobuf/issues/10075 is fixed :)

That one actually seems optimal given the situation, what do you think? We could even keep pinning on Windows so stuff works there too unless the pin is ignored.

**Pull request merged**